### DISCUSSION TITLE
Using upstash/redis to implement rate limiting on api routes of Next JS

### DISCUSSION THREAD

**Discussion by louremipsum** [vercel/next.js#62178] [Help]
### Summary

# Context
I wanted to implement rate limiting for a route in /api route in my next js app router app. Currently I am using upstash/redis [rate-limiting](https://github.com/upstash/ratelimit) and here are the [docs](https://upstash.com/docs/oss/sdks/ts/ratelimit/gettingstarted)  and I wanted to opt in for [Ephemeral Cache](https://github.com/upstash/ratelimit?tab=readme-ov-file#ephemeral-cache) so that it doesn't request redis everytime in case there is an extreme DDOS and to save time and money as Upstash is paid if too much data in storage

## Problem
It is written that i would have to make a `Map()` variable like `const cache = new Map(); // must be outside of your serverless function handler` and it is written that it should be outside my serverless function. so here is my code 

```typescript
// /api/myapi/route.ts

import cache from "@/cache/ratelimitCache";

const ratelimit = new Ratelimit({
  redis: redis,
  analytics: true,
  limiter: Ratelimit.slidingWindow(2, "5s"),
  prefix: "@upstash/ratelimit",
  ephemeralCache: cache,
});
```

```typescript
// /src/cache/ratelimitCache.ts

const cache = new Map();

export default cache;
```

So when i deploy my next js app on Vercel, will it be properly implemented and not just a static variable. I wanted to know the correct way to do it. 
Any help is appreciated!
Thank you

**Comment by a8m1r** [comment]
1. **Adjust your `ratelimitCache.ts` file**: Instead of directly exporting `cache`, create a function called `getCache` that returns the `cache` instance, ensuring it's the same across requests.

   ```typescript
   // /src/cache/ratelimitCache.ts

   let cache: Map<string, any> | null = null;

   export function getCache(): Map<string, any> {
     if (!cache) {
       cache = new Map();
     }
     return cache;
   }

2. **Update your rate limiting setup**:

   ```typescript
   // /api/myapi/route.ts

   import { getCache } from "@/cache/ratelimitCache";

   const cache = getCache();

   const ratelimit = new Ratelimit({
     redis: redis,
     analytics: true,
     limiter: Ratelimit.slidingWindow(2, "5s"),
     prefix: "@upstash/ratelimit",
     ephemeralCache: cache,
   });

**Reply by louremipsum** [reply]
I tested it and it is working on localhost(I tried to console log the cache variable but there was no output apart from 
```
cache-> Map(0) {}    // from ratelimitCache.ts
cache from route->  Map(0) {}           // /api/myroute/route.ts
```
. Still, when I deploy it, it says the rate limit is exceeded on every request no matter what I do and my Upstash Redis DB also has no keys named: `"@upstash/ratelimit"` as said in the prefix. This is my full code
```typescript
import { Ratelimit } from "@upstash/ratelimit";
import { NextRequest } from "next/server";
import redis from "./redis";
import { getCache } from "@/cache/ratelimitCache";
import { ipAddress } from "@vercel/edge";

export const runtime = "edge";

const cache = getCache();

const ratelimit = new Ratelimit({
  redis: redis,
  analytics: true,
  limiter: Ratelimit.slidingWindow(5, "3s"),
  prefix: "@upstash/ratelimit",
  ephemeralCache: cache,
});

export async function GET(request: NextRequest) {
  try {
    const id = ipAddress(request) || request.headers.get("x-forwarded-for");

    const limit = await ratelimit.limit(id ?? "anonymous");
    if (!limit.success) {
      return new Response("Rate limit exceeded", { status: 429 });
    }
    const secretKey = request.headers.get("x-secret-key");
    if (secretKey !== process.env.INTERNAL_SECRET_KEY) {
      return new Response("Unauthorized", { status: 403 });
    }
    const { searchParams } = new URL(request.url);
    const manageAPI = searchParams.get("manageAPI");
    if (manageAPI === null) {
      return new Response("No manageAPI parameter provided", { status: 400 });
    }
    let manageAPIBool: boolean = false;
    if (manageAPI === "true") {
      manageAPIBool = true;
    } else if (manageAPI === "false") {
      manageAPIBool = false;
    }
   // other code
  } catch (error) {
    return new Response("Nothing obtained", { status: 500 });
  }
}
```
I wanted to block if too many requests come from the same IP, tried to do it with `request.ip` as suggested [here](https://nextjs.org/docs/app/api-reference/functions/next-request#ip) but it was undefined in both development and after hosting on Vercel so I had to make it an edge function but it's still not working. Thank you for reading

**Comment by louremipsum** [comment]
Thank you for the swift reply, I will implement and test it to give you feedback if it's working properly in my code
